backend/analytics_server/mhq/service/code/sync/revert_pr_gitlab_sync.py

```diff
@@ -0,0 +1,192 @@
+import re
+from datetime import datetime
+from typing import Dict, List, Optional, Set
+from mhq.store.models.code.enums import (
+    PullRequestRevertPRMappingActorType,
+    PullRequestState,
+)
+from mhq.store.models.code.pull_requests import PullRequest, PullRequestRevertPRMapping
+from mhq.store.repos.code import CodeRepoService
+
+
+class RevertPRsGitlabSyncHandler:
+    def __init__(self, code_repo_service: CodeRepoService):
+        self.code_repo_service = code_repo_service
+
+    def __call__(self, *args, **kwargs):
+        return self.process_revert_prs(*args, **kwargs)
+
+    def process_revert_prs(
+        self, prs: List[PullRequest]
+    ) -> List[PullRequestRevertPRMapping]:
+        revert_prs: List[PullRequest] = []
+        original_prs: List[PullRequest] = []
+
+        for pr in prs:
+            pr_merge_commit_sha = (
+                self.get_revert_merge_commit_hash(pr.head_branch)
+                if pr.head_branch
+                else None
+            )
+            if pr_merge_commit_sha is None:
+                original_prs.append(pr)
+            else:
+                revert_prs.append(pr)
+
+        mappings_of_revert_prs = self._get_revert_pr_mapping_for_revert_prs(revert_prs)
+        mappings_of_original_prs = self._get_revert_pr_mapping_for_original_prs(
+            original_prs
+        )
+        revert_pr_mappings = set(mappings_of_original_prs + mappings_of_revert_prs)
+
+        return list(revert_pr_mappings)
+
+    def _get_revert_pr_mapping_for_original_prs(
+        self, prs: List[PullRequest]
+    ) -> List[PullRequestRevertPRMapping]:
+        """
+        This function takes a list of PRs and for each PR it tries to
+        find if that pr has been reverted and by which PR. It is done
+        by taking repo_id and the merge commit hash and searching for
+        the string 'revert-[merge-commit-hash]' in the head branch.
+        """
+
+        repo_ids: Set[str] = set()
+        repo_id_to_pr_merge_hash_to_id_map: Dict[str, Dict[str, str]] = {}
+        pr_merge_hash_match_strings: List[str] = []
+
+        for pr in prs:
+            if pr.state != PullRequestState.MERGED:
+                continue
+
+            pr_merge_hash_match_strings.append(f"revert-{pr.merge_commit_sha[:8]}")
```

Comment:
is this pattern for revert pr branch names always true ?

Reply:
yep